packages/back-end/src/routers/vercel-native-integration/vercel-native-integration.controller.ts

```diff
@@ -0,0 +1,531 @@
+    };
+
+const checkAuth = async <T extends string | "user">({
+  token,
+  type,
+}: {
+  token: string;
+  type: T;
+}): Promise<CheckedAuth<T> | { status: "error"; message: string }> => {
+  try {
+    const payload = await jwtVerify(token, await vercelJKWSKey);
+
+    if (type === "user")
+      return {
+        status: "authenticated",
+        authentication: userAuthenticationValidator.parse(payload),
+      } as CheckedAuth<T>;
+
+    if (type === "system")
+      return {
+        status: "authenticated",
+        authentication: systemAuthenticationValidator.parse(payload),
+      } as CheckedAuth<T>;
+
+    return { status: "error", message: `Unsupported authentication: ${type}` };
+  } catch (_err) {
+    return { status: "error", message: "Invalid credentials" };
+  }
+};
+
+const addOrganizationUser = (id: string, organization: OrganizationInterface) =>
+  updateOrganization(organization.id, {
+    members: [
+      ...organization.members,
+      {
+        id,
+        role: "admin",
+        dateCreated: new Date(),
+        limitAccessByEnvironment: false,
+        environments: [],
+      },
+    ],
+  });
+
+const findOrCreateUser = async ({
+  email,
+  organization,
+}: {
+  email: string;
+  organization: OrganizationInterface;
+}) => {
+  const existingUser = await getUserByEmail(email);
+
+  if (existingUser) {
+    if (!organization.members.find(({ id }) => existingUser.id === id))
+      await addOrganizationUser(existingUser.id, organization);
+    return existingUser;
+  }
+
+  const newUser = await createUser({
+    name: email.split("@")[0],
+    email,
+    password: crypto.randomBytes(18).toString("hex"),
+  });
+
+  await addOrganizationUser(newUser.id, organization);
+
+  return newUser;
+};
+
+const getContext = async ({
+  installationId,
+  resourceId,
+  userEmail,
+  res,
+}: {
+  installationId: string;
+  resourceId?: string;
+  userEmail?: string;
+  res: Response;
+}) => {
+  const failed = (status: number, reason?: string) => {
+    if (reason) res.status(status).send(reason);
+    else res.sendStatus(status);
+
+    throw new Error("Authentication failed");
+  };
+
+  const integration = await findVercelInstallationByInstallationId(
+    installationId
+  );
+
+  if (!integration) return failed(400, "Invalid installation!");
+
+  const resource = (() => {
+    if (!resourceId) return;
+
+    const resource = integration.resources.find(({ id }) => id === resourceId);
+
+    if (!resource) {
+      res.status(400).send("Invalid request!");
+      throw new Error(`Invalid resourceId: ${resourceId}`);
+    }
+
+    return resource;
+  })();
+
+  const organizationId = resource?.organizationId || integration.organization;
+  const org = await findOrganizationById(organizationId);
+
+  if (!org) return failed(400, "Invalid installation!");
+
+  const user = await findOrCreateUser({
+    email: userEmail || org.ownerEmail,
+    organization: org,
+  });
```

Comment:
One potential issue here: if the user was just added to the org, this local `org` object will not have the user in the `members` array yet.  That will likely affect the permissions available to the context object.